data/store/mongo/mongo_data.go

```diff
@@ -0,0 +1,1042 @@
+	}
+	return nil
+}
+
+func validateAndTranslateSelectors(selectors *data.Selectors) (bson.M, error) {
+	if selectors == nil {
+		return bson.M{}, nil
+	} else if err := structureValidator.New().Validate(selectors); err != nil {
+		return nil, errors.Wrap(err, "selectors is invalid")
+	}
+
+	var selectorIDs []string
+	var selectorOriginIDs []string
+	for _, selector := range *selectors {
+		if selector != nil {
+			if selector.ID != nil {
+				selectorIDs = append(selectorIDs, *selector.ID)
+			} else if selector.Origin != nil && selector.Origin.ID != nil {
+				selectorOriginIDs = append(selectorOriginIDs, *selector.Origin.ID)
+			}
+		}
+	}
+
+	selector := bson.M{}
+	if len(selectorIDs) > 0 && len(selectorOriginIDs) > 0 {
+		selector["$or"] = []bson.M{
+			{"id": bson.M{"$in": selectorIDs}},
+			{"origin.id": bson.M{"$in": selectorOriginIDs}},
+		}
+	} else if len(selectorIDs) > 0 {
+		selector["id"] = bson.M{"$in": selectorIDs}
+	} else if len(selectorOriginIDs) > 0 {
+		selector["origin.id"] = bson.M{"$in": selectorOriginIDs}
+	}
+
+	if len(selector) == 0 {
+		return nil, errors.New("selectors is invalid")
+	}
+
+	return selector, nil
+}
+
+func (d *DataRepository) GetCGMDataRange(ctx context.Context, id string, startTime time.Time, endTime time.Time) ([]*continuous.Continuous, error) {
+	var dataSets []*continuous.Continuous
+	selector := bson.M{
+		"_active": true,
+		"_userId": id,
+		"type":    "cbg",
+		// we currently pull all so we can do in-range and avg in 1 pass, maybe bad idea
+		//"value": bson.M{"$gt": 3.9, "$lt": 10},
+		// NOTE: redundant query for migration of time field
+		"$or": bson.A{
+			bson.M{"time": bson.M{"$gte": startTime.Format(time.RFC3339Nano),
+				"$lte": endTime.Format(time.RFC3339Nano)}},
+			bson.M{"time": bson.M{"$gte": startTime,
+				"$lte": endTime}},
+		},
+	}
+
+	cursor, err := d.Find(ctx, selector)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get in-range values")
+	}
+
+	if err = cursor.All(ctx, &dataSets); err != nil {
+		return nil, errors.Wrap(err, "unable to decode data sets for user by id")
+	}
+
+	return dataSets, nil
+}
+
+func (d *DataRepository) GetLastUpdatedForUser(ctx context.Context, id string) (summary.UserLastUpdated, error) {
+	var status summary.UserLastUpdated
+
+	if ctx == nil {
+		return status, errors.New("context is missing")
+	}
+
+	if id == "" {
+		return status, errors.New("id is missing")
+	}
+
+	var dataSet []*continuous.Continuous
+	selector := bson.M{
+		"_active": true,
+		"_userId": id,
+		"type":    "cbg",
+	}
+
+	findOptions := options.Find()
+	findOptions.SetSort(bson.D{{Key: "time", Value: -1}})
+	findOptions.SetLimit(1)
+
+	cursor, err := d.Find(ctx, selector, findOptions)
+
+	if err != nil {
+		return status, errors.Wrap(err, "unable to get last cbg date")
+	}
+
+	if err = cursor.All(ctx, &dataSet); err != nil {
+		return status, errors.Wrap(err, "unable to decode last cbg date")
+	}
+
+	if dataSet != nil {
```

Comment:
You should check the if `len(dataSet) > 0` instead of checking if the data set is nil. I think it can never be `nil`.